Use TextIndicator instead of TextIndicatorData in createDragImageForLink.

megangardner opened the pull request:
#### cac26de4a923efa653f577fed297ed3e2bed8430
<pre>
Use TextIndicator instead of TextIndicatorData in createDragImageForLink.
<a href="https://bugs.webkit.org/show_bug.cgi?id=294473">https://bugs.webkit.org/show_bug.cgi?id=294473</a>
<a href="https://rdar.apple.com/problem/153335624">rdar://problem/153335624</a>

Reviewed by Abrar Rahman Protyasha.

Continuing to move off of TextIndicatorData.
Also, refactoring this code to return everything
instead of using out variables.

* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::startDrag):
* Source/WebCore/platform/DragImage.cpp:
(WebCore::createDragImageForLink):
* Source/WebCore/platform/DragImage.h:
* Source/WebCore/platform/cocoa/DragImageCocoa.mm:
(WebCore::createDragImageForLink):
* Source/WebCore/platform/ios/DragImageIOS.mm:
(WebCore::createDragImageForLink):

Canonical link: <a href="https://commits.webkit.org/296242@main">https://commits.webkit.org/296242@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3e118d43f17e809e0217fe72795901ce7945d729

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107926 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27591 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18010 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113137 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58448 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109889 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28286 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36140 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81917 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110874 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22412 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97236 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62348 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21824 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15369 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57890 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91763 "Passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15435 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 11 flakes 3 failures; Uploaded test results; 10 flakes 3 failures; Running compile-webkit-without-change") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116264 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34998 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25753 "Found 1 new test failure: fast/webgpu/nocrash/fuzz-291180.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90950 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35374 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93514 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90744 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23116 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35647 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13401 "Passed tests") | [⏳ 🛠 playstation ](https://ews-build.webkit.org/#/builders/PlayStation-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34896 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40448 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Checked out pull request; Found 74135 issues") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34639 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37999 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36300 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b83c89ec68950b2b45076aee80e661a81a9a86f4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46732

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b83c89ec68950b2b45076aee80e661a81a9a86f4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46732

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/785ef42b5b3220ff97856054e851564621cc8e82. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46732

aprotyas commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -1190,14 +1190,14 @@ bool DragController::startDrag(LocalFrame& src, const DragState& state, OptionSe
-            TextIndicatorData textIndicator;
-            dragImage = DragImage { createDragImageForLink(element, linkURL, textContentWithSimplifiedWhiteSpace, textIndicator, m_page->deviceScaleFactor()) };
+            auto [dragImageRef, textIndicator] = createDragImageForLink(element, linkURL, textContentWithSimplifiedWhiteSpace, m_page->deviceScaleFactor());
+            dragImage = DragImage { dragImageRef };
```
While `DragImage` currently does not support move semantics, it doesn't hurt to:
```suggestion
            dragImage = DragImage { WTFMove(dragImageRef) };
```

Right now the `WTFMove(...)` will be a no-op, but once we have this constructor, this will move the text indicator:

```cpp
DragImage(RefPtr<TextIndicator>&&, std::optional<Path>&&);
```

aprotyas commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -1190,14 +1190,14 @@ bool DragController::startDrag(LocalFrame& src, const DragState& state, OptionSe
-                if (textIndicator.contentImage)
-                    dragImage.setTextIndicator(TextIndicator::create(textIndicator));
+                if (textIndicator && textIndicator->contentImage())
+                    dragImage.setTextIndicator(textIndicator);
```
Same comment here:

```suggestion
                    dragImage.setTextIndicator(WTFMove(textIndicator));
```

... if we have `void setTextIndicator(RefPtr<TextIndicator>&&)`

aprotyas commented on Source/WebCore/platform/win/cairo/DragImageWinCairo.cpp:
```diff
@@ -327,7 +327,7 @@ DragImageRef createDragImageForLink(Element&, URL& url, const String& inLabel, T
 
     deallocContext(contextRef);
-    return image.leak();
+    return  { image.leak(), nullprt };
```
Typo here
```suggestion
    return  { image.leak(), nullptr };
```

rr-codes replied:
(and there's an extra space before the { I think)

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2db2caa4777f8a4ac52efcba90d76bad7ce8a510. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46732

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3e118d43f17e809e0217fe72795901ce7945d729. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46732

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3e118d43f17e809e0217fe72795901ce7945d729. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46732

megangardner committed:
Use TextIndicator instead of TextIndicatorData in createDragImageForLink.
https://bugs.webkit.org/show_bug.cgi?id=294473
rdar://problem/153335624

Reviewed by Abrar Rahman Protyasha.

Continuing to move off of TextIndicatorData.
Also, refactoring this code to return everything
instead of using out variables.

* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::startDrag):
* Source/WebCore/platform/DragImage.cpp:
(WebCore::createDragImageForLink):
* Source/WebCore/platform/DragImage.h:
* Source/WebCore/platform/cocoa/DragImageCocoa.mm:
(WebCore::createDragImageForLink):
* Source/WebCore/platform/ios/DragImageIOS.mm:
(WebCore::createDragImageForLink):

Canonical link: https://commits.webkit.org/296242@main

webkit-commit-queue commented:
Committed 296242@main (cac26de4a923): <https://commits.webkit.org/296242@main>

Reviewed commits have been landed. Closing PR #46732 and removing active labels.
